distributed/semaphore.py

```diff
@@ -138,37 +173,44 @@ def release(self, comm=None, name=None, client=None, identifier=None):
                 return
             if isinstance(name, list):
                 name = tuple(name)
-            if name in self.leases and identifier in self.leases[name]:
-                self._release_value(name, client, identifier)
+            if name in self.leases and lease_id in self.leases[name]:
+                self._release_value(name, lease_id)
             else:
                 raise ValueError(
                     f"Tried to release semaphore but it was already released: "
-                    f"client={client}, name={name}, identifier={identifier}"
+                    f"name={name}, lease_id={lease_id}"
                 )
 
-    def _release_value(self, name, client, identifier):
+    def _release_value(self, name, lease_id):
+        logger.info("Releasing %s for %s", lease_id, name)
         # Everything needs to be atomic here.
-        self.leases_per_client[client][name].remove(identifier)
-        self.leases[name].remove(identifier)
+        del self.leases[name][lease_id]
         self.events[name].set()
 
-    def _release_client(self, client):
-        semaphore_names = list(self.leases_per_client[client])
-        for name in semaphore_names:
-            ids = list(self.leases_per_client[client][name])
-            for _id in list(ids):
-                self._release_value(name=name, client=client, identifier=_id)
-        del self.leases_per_client[client]
-
     def _validate_leases(self):
         if not self._validation_running:
             self._validation_running = True
-            known_clients_with_leases = set(self.leases_per_client.keys())
-            scheduler_clients = set(self.scheduler.clients.keys())
-            for dead_client in known_clients_with_leases - scheduler_clients:
-                self._release_client(dead_client)
-            else:
-                self._validation_running = False
+            now = time()
+            semaphore_names = list(self.leases.keys())
```

Comment:
I'm curious, is there a reason (preserving determinism?) for casting `dict_keys -> list` here (and doing the same for `ids` a few lines below)?

Reply:
Otherwise, this will raise a `dictionary changed size during iteration` exception when a lease is removed